File: app/src/main/java/dev/ragnarok/fenrir/adapter/DocsAsImagesAdapter.java
package dev.ragnarok.fenrir.adapter;

import static dev.ragnarok.fenrir.util.Objects.nonNull;
import static dev.ragnarok.fenrir.util.Utils.nonEmpty;

import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.EventListener;
import java.util.List;

import dev.ragnarok.fenrir.Constants;
import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.adapter.base.RecyclerBindableAdapter;
import dev.ragnarok.fenrir.model.Document;
import dev.ragnarok.fenrir.model.PhotoSize;
import dev.ragnarok.fenrir.picasso.PicassoInstance;

public class DocsAsImagesAdapter extends RecyclerBindableAdapter<Document, DocsAsImagesAdapter.DocViewHolder> {

    private ActionListener mActionListener;

    public DocsAsImagesAdapter(List<Document> data) {
        super(data);
    }

    public void setData(List<Document> data) {
        setItems(data);
    }

    public void setActionListener(ActionListener listener) {
        mActionListener = listener;
    }

    @Override
    protected void onBindItemViewHolder(DocViewHolder holder, int position, int type) {
        Document item = getItem(position);

        holder.title.setText(item.getTitle());

        String previewUrl = item.getPreviewWithSize(PhotoSize.Q, false);
        boolean withImage = nonEmpty(previewUrl);

        if (withImage) {
            PicassoInstance.with()
                    .load(previewUrl)
                    .tag(Constants.PICASSO_TAG)
                    .into(holder.image);
        } else {
            PicassoInstance.with()
                    .cancelRequest(holder.image);
        }

        holder.itemView.setOnClickListener(v -> {
            if (nonNull(mActionListener)) {
                mActionListener.onDocClick(holder.getBindingAdapterPosition(), item);
            }
        });

        holder.itemView.setOnLongClickListener(v -> nonNull(mActionListener)
                && mActionListener.onDocLongClick(holder.getBindingAdapterPosition(), item));
    }

    @Override
    protected DocViewHolder viewHolder(View view, int type) {
        return new DocViewHolder(view);
    }

    @Override
    protected int layoutId(int type) {
        return R.layout.item_doc_as_image;
    }

    public interface ActionListener extends EventListener {
        void onDocClick(int index, @NonNull Document doc);

        boolean onDocLongClick(int index, @NonNull Document doc);
    }

    static class DocViewHolder extends RecyclerView.ViewHolder {

        final ImageView image;
        final TextView title;

        DocViewHolder(View root) {
            super(root);
            image = root.findViewById(R.id.image);
            title = root.findViewById(R.id.title);
        }
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/fragment/docs/DocsAsImagesAdapter.kt
package dev.ragnarok.fenrir.fragment.docs

import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import dev.ragnarok.fenrir.Constants
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.fragment.base.RecyclerBindableAdapter
import dev.ragnarok.fenrir.model.Document
import dev.ragnarok.fenrir.model.PhotoSize
import dev.ragnarok.fenrir.nonNullNoEmpty
import dev.ragnarok.fenrir.picasso.PicassoInstance.Companion.with

class DocsAsImagesAdapter(data: MutableList<Document>) :
    RecyclerBindableAdapter<Document, DocsAsImagesAdapter.DocViewHolder>(data) {
    private var mActionListener: ActionListener? = null
    fun setData(data: MutableList<Document>) {
        setItems(data)
    }

    fun setActionListener(listener: ActionListener?) {
        mActionListener = listener
    }

    override fun onBindItemViewHolder(viewHolder: DocViewHolder, position: Int, type: Int) {
        val item = getItem(position)
        viewHolder.title.text = item.title
        val previewUrl = item.getPreviewWithSize(PhotoSize.Q, false)
        val withImage = previewUrl.nonNullNoEmpty()
        if (withImage) {
            with()
                .load(previewUrl)
                .tag(Constants.PICASSO_TAG)
                .into(viewHolder.image)
        } else {
            with()
                .cancelRequest(viewHolder.image)
        }
        viewHolder.itemView.setOnClickListener {
            mActionListener?.onDocClick(getItemRawPosition(viewHolder.bindingAdapterPosition), item)
        }
        viewHolder.itemView.setOnLongClickListener {
            mActionListener?.onDocLongClick(
                getItemRawPosition(viewHolder.bindingAdapterPosition),
                item
            ) == true
        }
    }

    override fun viewHolder(view: View, type: Int): DocViewHolder {
        return DocViewHolder(view)
    }

    override fun layoutId(type: Int): Int {
        return R.layout.item_doc_as_image
    }

    interface ActionListener {
        fun onDocClick(index: Int, doc: Document)
        fun onDocLongClick(index: Int, doc: Document): Boolean
    }

    class DocViewHolder(root: View) : RecyclerView.ViewHolder(root) {
        val image: ImageView = root.findViewById(R.id.image)
        val title: TextView = root.findViewById(R.id.title)
    }
}